src/auth/impersonated.ts

```diff
@@ -0,0 +1,194 @@
+   * @param lifetime number of seconds the delegated credential should be
+   * valid for (up to 3600).
+   * @param endpoint api endpoint override.
+   */
+  constructor(options: ImpersonatedOptions = {}) {
+    super(options);
+    this.credentials = {
+      expiry_date: 1,
+      refresh_token: 'impersonated-placeholder',
+    };
+    this.sourceClient = options.sourceClient || new OAuth2Client();
+    this.targetPrincipal = options.targetPrincipal || '';
+    this.delegates = options.delegates || [];
+    this.targetScopes = options.targetScopes || [];
+    this.lifetime = options.lifetime || 3600;
+    this.endpoint = options.endpoint || 'https://iamcredentials.googleapis.com';
+  }
+
+  /**
+   * Refreshes the access token.
+   * @param refreshToken Unused parameter
+   */
+  protected async refreshToken(
+    refreshToken?: string | null
+  ): Promise<GetTokenResponse> {
+    const iat = Date.now();
+    if (this.credentials.expiry_date) {
+      if (this.credentials.expiry_date <= iat) {
+        try {
+          await this.sourceClient.getAccessToken();
+          const name = 'projects/-/serviceAccounts/' + this.targetPrincipal;
+          const u = `${this.endpoint}/v1/${name}:generateAccessToken`;
+          const body = {
+            delegates: this.delegates,
+            scope: this.targetScopes,
+            lifetime: this.lifetime + 's',
+          };
+          const res = await this.sourceClient.request({
+            url: u,
+            data: body,
+            method: 'POST',
+          });
+          const tokenResponse = res.data as TokenResponse;
+          this.credentials.access_token = tokenResponse.accessToken;
+          this.credentials.expiry_date =
+            Date.parse(tokenResponse.expireTime) / 1000;
+          return {
+            tokens: this.credentials,
+            res,
+          };
+        } catch (error) {
+          if (error.status === 403) {
+            if (error.message === 'The caller does not have permission') {
+              throw new Error(
+                'Error: Unable to impersonate: sourceCredential lacks IAM Token Creator role on targetCredential'
+              );
+            }
+            if (
+              error.message ===
+              'Request had insufficient authentication scopes.'
+            ) {
+              throw new Error(
+                'Error: Unable to impersonate: sourceCredential lacks cloud-platform or IAM scope'
+              );
+            }
+          }
+          throw new Error('Error: Unable to impersonate: ' + error);
+        }
+      }
+      return {tokens: this.credentials, res: null};
+    }
+    throw new Error('Error: Root credentials.expiry_date not set ');
+  }
+
+  protected requestAsync<T>(
+    opts: GaxiosOptions,
+    retry = false
+  ): GaxiosPromise<T> {
+    try {
+      return super.requestAsync<T>(opts, retry);
+    } catch (err) {
+      let helpfulMessage = null;
+      if (err.status === 403) {
+        helpfulMessage =
```

Comment:
Likewise, perhaps we can get away with checking for just the `403` and `404` case?